After initial EDA I figured out that the of the author provided "Bird_Species.csv" misses some classes and files, but the classes are present and there are also images for these classes. Therefore, I first created a new cleaned version of this csv file and use this one for my project. <br> <br>
This step is not required, because I could simply use the "flow_from_directory" function from tensorflow. But I not just want to train the model blindly on a dataset without knowing the distributions! A good machine learning engineer should always be in full awarness about his dataset and the distributions.

# Data Import

In [1]:
import sys
import os
import numpy as np
import pandas as pd
import csv
import glob
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

plt.rcParams['figure.figsize'] = [8,8]
#sns.set_style("whitegrid")

In [2]:
# get dataframe with all classes
df_classes = pd.read_csv(r"data\class_dict.csv")
df_classes.head()

class_index                  class  height  width  scale by
0            0  AFRICAN CROWNED CRANE     224    224         1
1            1      AFRICAN FIREFINCH     224    224         1
2            2              ALBATROSS     224    224         1
3            3   ALEXANDRINE PARAKEET     224    224         1
4            4        AMERICAN AVOCET     224    224         1

In [3]:
# check the number of classes -> should be 300
print(f"There are {df_classes.shape[0]} different classes in this dataset.")

There are 300 different classes in this dataset.


In [4]:
# read in original dataset csv file to show that some classes are missing
df = pd.read_csv(r"data\Bird_Species.csv")
df.head()

filepaths                 labels data set
0  train/AFRICAN CROWNED CRANE/001.jpg  AFRICAN CROWNED CRANE    train
1  train/AFRICAN CROWNED CRANE/002.jpg  AFRICAN CROWNED CRANE    train
2  train/AFRICAN CROWNED CRANE/003.jpg  AFRICAN CROWNED CRANE    train
3  train/AFRICAN CROWNED CRANE/004.jpg  AFRICAN CROWNED CRANE    train
4  train/AFRICAN CROWNED CRANE/005.jpg  AFRICAN CROWNED CRANE    train

In [5]:
num_unique_classes = df["labels"].unique().shape[0]
print(f"There are {num_unique_classes} unique class labels in the dataset.")

There are 285 unique class labels in the dataset.


As you can see: This dataset only contains 285 classes instead of 300, but in the directories there are 300 subdirectories with images in each of them. Therefore, let's create a new cleaned version of this dataset.

In [6]:
import glob
def get_list_of_files(folder):
    return glob.glob(os.path.join(folder, "*.jpg"))

In [11]:
import csv

foldernames = ["train", "valid", "test"]
path_csv = r"data\Bird_Species_cleaned.csv"
columns = ["filepaths", "labels", "data set"]

# write header
with open(path_csv, "w", newline="") as csv_file:
    csv_writer = csv.writer(csv_file, quoting=csv.QUOTE_NONE)
    csv_writer.writerow(columns)

for foldername in foldernames:
    # loop over all classes (subdirectories) and get all filenames located in the classes
    for class_folder in df_classes["class"]:
        list_files = get_list_of_files(os.path.join("data", foldername, class_folder))

        with open(path_csv, "a", newline="") as csv_file:
            csv_writer = csv.writer(csv_file, quoting=csv.QUOTE_NONE)
            for filepath in list_files:
                # first: remove data from filepath
                filepath = os.path.sep.join(filepath.split(os.path.sep)[1:])
                csv_writer.writerow([filepath, class_folder, foldername])     